tools/codegen/gen.py

```diff
@@ -460,26 +460,92 @@ def compute_meta_function_declaration(g: NativeFunctionsGroup) -> Optional[str]:
 
         if precomputed:
+            # Generate the template declaration with one bool parameter for each
+            # precomputed element. Each parameter is true if the corresponding (in
+            # terms of position) precomputed element has been set.
+            precomputed_elements = [elem for replace_list in precomputed.replace.values() for elem in replace_list]
+            precomputed_template_parameters = [chr(ord('A') + i) for i in range(len(precomputed_elements))]
+            precomputed_template_params_str = ", ".join(f"bool {param} = false" for param in precomputed_template_parameters)
+            precompute_template_decl = f"template <{precomputed_template_params_str}>"
+
             # Generate a string containing declarations of all precomputed elements.
-            precomputed_elements_with_cpp_types = (
+            precomputed_elements_with_cpp_types = [
                 structured.argument_type(elem, binds=elem.name)
-                for replace_list in precomputed.replace.values() for elem in replace_list
-            )
-            precomputed_elements = ";\n".join(
+                for elem in precomputed_elements
+            ]
+
+            precomputed_elements_decl = ";\n".join(
                 f"{elem.cpp_type(strip_ref=True)} {elem.name}" for elem in precomputed_elements_with_cpp_types
             )
+
+            # Generate "setter" methods for each precomputed element. Each method will return
+            # a new instance of precompute_out with the template parameter that corresponds to
+            # the member set by the method to true (to indicate that it has been set).
+            setter_methods = []
+            for i, elem in enumerate(precomputed_elements):
+                # Generate the signature. The return type will be the same
+                # as the type of `this` but with the template parameter
+                # corresponding to the element set by this method set to true.
+                # The assert generated below will ensure that this template
+                # parameter is false on the type of `this`.
+                return_ty_templates = ", ".join(
+                    precomputed_template_parameters[:i] + ["true"] + precomputed_template_parameters[i + 1:]
+                )
+                return_ty = f"precompute_out<{return_ty_templates}>"
+                elem_cpp_ty = precomputed_elements_with_cpp_types[i].cpp_type(strip_ref=True)
+                signature = f"{return_ty} set_{elem.name}({elem_cpp_ty} value)"
+
+                # Generate an assert which checks that the
+                # template parameter corresponding to the precomputed
+                # element that is set by this method is false on the
+                # class corresponding to the object that `this` points to.
+                # This ensures that each element can be set only once.
+                assert_stmt = f"TORCH_INTERNAL_ASSERT({precomputed_template_parameters[i]} == false);"
```

Comment:
I tried to use `std::enable_if` to disable the corresponding method altogether when this condition is not met, but that led to not being able to instantiate `precompute_out<true, true, ..., true>` (i.e. the final instance will all members set). Hopefully I did something dumb and there is a way to convert this assert into a a compile time error.

Reply:
Oh, what about `static_assert`? :O